### PR TITLE
feat: Add allowTrailingCommas option for JSONC

### LINKED ISSUE #41: Feature: Dangling commas in JSONC
### Environment

ESLint version: 9.12.0
@eslint/json version: 0.5.0
Node version: 22.9.0
npm version: 10.9.0
Operating System: MacOS 15.0.1

### What problem do you want to solve?

The json plugin parses JSONC files with @humanwhocodes/momoa.  In version 3.3.0 of momoa, there is now the option to allow trailing commas in JSONC files, using the `allowTrailingCommas: true` option.  See the [decision](https://github.com/humanwhocodes/momoa/blob/main/decisions/002-json-jsonc-trailing-commas.md) for more information.

As a new feature in @eslint/json, it would be nice to either:
a) Turn on this option by default for JSONC
b) Allow setting this in the languageOptions for JSONC, with default to true
c) Allow setting this in the languageOptions for JSONC, with default to false

Because many common JSONC files (all of the vscode config files and the tsconfig.json file) allow trailing commas, a-c are my order of preference.  However, since the Microsoft JSONC parser defaults to false, it is completely reasonable to pick c.

[link](https://github.com/humanwhocodes/momoa/issues/135) to momoa issue.

### What do you think is the correct solution?

Turn on allowTrailingCommas in the call to momoa.

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

-   [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

Add support for trailing commas to the JSONC language.

#### What changes did you make? (Give an overview)

This pull request introduces support for allowing trailing commas in JSONC files and includes several updates to the codebase and documentation to reflect this new feature. The most important changes include adding a new option to enable trailing commas, updating the `JSONLanguage` class to validate and handle this option, and adding tests to ensure the feature works as expected.

### New Feature: Allow Trailing Commas in JSONC

* **README.md**: Added a section explaining how to enable trailing commas in JSONC files using the `allowTrailingCommas` language option. 

### Code Updates

* **`src/languages/json-language.js`**:
  * Added a new typedef for `JSONLanguageOptions` to include the `allowTrailingCommas` property. 
  * Updated the `validateLanguageOptions` method to check for the `allowTrailingCommas` option and ensure it is only used in JSONC mode.
  * Modified the `parse` method to accept and handle the `allowTrailingCommas` option. 

### Dependency Update

* **package.json**: Updated the `@humanwhocodes/momoa` dependency to version `^3.3.0`. 

### Tests

* **`tests/languages/json-language.test.js`**:
  * Added tests for the `validateLanguageOptions` method to ensure it correctly handles the `allowTrailingCommas` option. 
  * Added tests for the `parse` method to verify that trailing commas are correctly parsed or rejected based on the `allowTrailingCommas` option.

#### Related Issues

fixes #41

#### Is there anything you'd like reviewers to focus on?

Right now, `validateLanguageOptions()` throws an error when it finds `allowTrailingCommas` in JSON mode (where it cannot be enabled) and JSON5 mode (where it cannot be disabled). Would it be better to check the value of `allowTrailingCommas` before deciding to throw an error? So in JSON mode it would only throw if `allowTrailingCommas` is `true` and in JSON5 mode it would only throw an error if `allowTrailingCommas` is `false`?

### CODE REVIEW

**Comment by mdjermanovic** [comment]
> Right now, `validateLanguageOptions()` throws an error when it finds `allowTrailingCommas` in JSON mode (where it cannot be enabled) and JSON5 mode (where it cannot be disabled). Would it be better to check the value of `allowTrailingCommas` before deciding to throw an error? So in JSON mode it would only throw if `allowTrailingCommas` is `true` and in JSON5 mode it would only throw an error if `allowTrailingCommas` is `false`?

I think it's fine to throw an error when this option is used in a mode where it isn't applicable, regardless of the value.

File: README.md [hunk 1]
```diff
@@ -185,6 +185,37 @@ In JSONC and JSON5 files, you can also use [rule configurations comments](https:
...
+		...json.configs.recommended,
+	},
+
+	// lint JSONC files and allow trailing commas
+	{
+		files: ["**/tsconfig.json", ".vscode/*.json"],
+		language: "json/jsonc",
+		languageOptions: {
+			allowTrailingCommas: true
+		}
```
Comment (mdjermanovic) [inline README.md:210]:
```suggestion
		},
```

**Comment by fasttime** [comment]
@nzakas There's a check failing because of a formatting problem reported by Prettier. Can you take a look?

**Comment by nzakas** [comment]
Hmm, I'm not seeing anything locally. 🤔

**Comment by nzakas** [comment]
Rebasing worked. :+1:

**Comment by mdjermanovic** [comment]
Sorry, I left a suggestion to fix the prettier error last week, by apparently forgot to click the submit review button.

File: README.md [hunk 2]
```diff
@@ -185,6 +185,37 @@ In JSONC and JSON5 files, you can also use [rule configurations comments](https:
...
+		...json.configs.recommended,
+	},
+
+];
```
Comment (mdjermanovic) [inline README.md:214]:
```suggestion
export default [
	// lint JSONC files
	{
		files: ["**/*.jsonc"],
		language: "json/jsonc",
		...json.configs.recommended,
	},

	// lint JSONC files and allow trailing commas
	{
		files: ["**/tsconfig.json", ".vscode/*.json"],
		language: "json/jsonc",
		languageOptions: {
			allowTrailingCommas: true
		},
		...json.configs.recommended,
	},
];
```

Comment (mdjermanovic) [reply README.md:214]: Missing comma caused a syntax error, so prettier was ignoring this code block.

File: README.md [hunk 3]
```diff
@@ -185,6 +185,35 @@ In JSONC and JSON5 files, you can also use [rule configurations comments](https:
...
+		language: "json/jsonc",
+		...json.configs.recommended,
+	},
+
+	// lint JSONC files and allow trailing commas
+	{
+		files: ["**/tsconfig.json", ".vscode/*.json"],
+		language: "json/jsonc",
+		languageOptions: {
+			allowTrailingCommas: true
```
Comment (mdjermanovic) [inline README.md:208]:
```suggestion
			allowTrailingCommas: true,
```

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Comment by fasttime** [comment]
> hi, when I add allowTrailingCommas options, typescript system will error that allowTrailingCommas isn't in the languageoptions, how to solve this?

@charliez0 Can you [create a bug report](https://github.com/eslint/json/issues/new?template=bug-report.yml) and show the code where the error occurs? Thanks!

**Pull request merged**